app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -204,26 +204,52 @@ private void handleTagNamesMessage(final TagNamesMessage message) {
             if (backfillPeriods > 0) {
+                final String sourceMetricName = getSourceMetricName(metricName, period);
+                final String rollupMetricName = getDestinationMetricName(metricName, period);
                 Patterns.pipe(
-                        fetchLastDataPoint(metricName + period.getSuffix(), period, backfillPeriods)
-                                .handle((response, failure) -> {
-                                    final String baseMetricName = "rollup/generator/last_data_point_"
-                                            + period.name().toLowerCase(Locale.getDefault());
-                                    _metrics.recordCounter(baseMetricName + "/success", failure == null ? 1 : 0);
-                                    _metrics.recordTimer(
-                                            baseMetricName + "/latency",
-                                            System.nanoTime() - startTime,
-                                            Optional.of(Units.NANOSECOND));
-                                    return buildLastDataPointResponse(metricName, period, message.getTagNames(), response, failure);
-                                }), getContext().dispatcher())
-                        .to(getSelf());
+                    fetchLastDataPoint(sourceMetricName, rollupMetricName, period, backfillPeriods)
+                        .handle((response, failure) -> {
+                            final String baseMetricName = "rollup/generator/last_data_point_"
+                                    + period.name().toLowerCase(Locale.getDefault());
+                            _metrics.recordCounter(baseMetricName + "/success", failure == null ? 1 : 0);
+                            _metrics.recordTimer(
+                                 baseMetricName + "/latency",
+                                 System.nanoTime() - startTime,
+                                Optional.of(Units.NANOSECOND)
+                            );
+                            return buildLastDataPointResponse(
+                                sourceMetricName,
+                                rollupMetricName,
+                                period,
+                                message.getTagNames(),
+                                response,
+                                failure
+                            );
+                        }),
+                    getContext().dispatcher()
+                ).to(getSelf());
             }
         }
     }
 
-    private void handleLastDataPointMessage(final LastDataPointMessage message) {
+    private String getSourceMetricName(final String metricName, final RollupPeriod period) {
+        Optional<RollupPeriod> sourcePeriod = period.nextSmallest();
+        while (sourcePeriod.isPresent() && !sourcePeriod.map(p -> _maxBackFillByPeriod.getOrDefault(p, 0) > 0).orElse(true)) {
+            sourcePeriod = sourcePeriod.flatMap(RollupPeriod::nextSmallest);
```

Comment:
Is "next smallest" going to try all values less than the roll-up or only all _possible_ values less than the roll-up?

It will only work if it tries all the values, which for now is probably ok, but please leave a comment that this is potentially inefficient and also potentially incorrect.

Specifically, this is a GCD type problem which is a tree. For example, 1H can be computed from 1M, 15M, 30M roll-ups, but cannot be computed from 25M and 45M roll-ups. So walking down from 1H (60M) would walk consider impossible values like 45M and 25M and if not recognized as impossible could try to roll-up from them! (which may result in bad data)

The generalization of this is:
* Each _possible_ roll-up declares its ancestors (e.g. in the enum):
```
1H => [30M, 15M, 1M]
45M => [15M, 1M]
30M => [15M, 1M]
15M => [1M]
```
* Then the specified roll-up configuration (e.g. 1H, 30M) must look at the enumeration to determine what are possible ancestors for each specified roll-up. In this case those would be:
```
1H = [30M, 1M]
30M = [1M]
```
(the 15M possibility drops off because it's not configured!).

Anyway, I don't expect this to be solved now. However, @speezepearson should be aware of it and @cwbriones can you file an issue and add a comment to the roll-up enum describing this to remind us not to a possible roll-up that exposes this behavior.

Reply:
I agree with the generalization but I also wrote this with the expectation that we guarantee each period size is divisible by each smaller period. So if we did have additional rollups for example I'd only expect something like `15m, 30m, 3h, 6h, 12h`. Of course, if that property does not hold then of course this will not work but I really don't see the value-add in complicating it for ourselves by doing it otherwise.

Regardless, I'll leave a comment in `RollupPeriod` since this code will not need to change even if we did add additional periods like this.

Reply:
Totally fair.

Assuming we have such well behaved supported roll-ups but someone only enables 30min and 1h? Does it know not to try to use 15min?

Reply:
Yes.

Maybe I should leave a comment, but this will walk backwards down "nextSmallest" until it finds the first one that's enabled.

So it would go - 
1. computing hourly
    * next smallest is 30m. it's enabled. use 30m
2. computing 30m, 
    * next smallest is 15m. it's disabled.
    * next smallest is 1m. use 1m.

Reply:
[I included a test for this in the most recent push](https://github.com/ArpNetworking/metrics-portal/blob/54f1d26dc31db25b2acc3242ab17e4ce572fe3c7/test/java/com/arpnetworking/rollups/RollupGeneratorTest.java#L234)

Reply:
I've thought about it more and I've found a case that hits the issue you've discussed.

Consider series with period `30m 15m 10m 1m`, and suppose we want to compute the `30m` rollup. All periods are enabled *except 15m*.

The current code will walk: 30m -> 15m (disabled) -> 5m and choose the `5m` rollup to use as the source instead of the `10m` because 10m does not evenly divide 15. This is non-optimal since it does divide 30.

This behavior doesn't occur in the `1h 45m 30m 15m` example.